packages/owl-bot/src/copy-code.ts

```diff
@@ -362,3 +368,65 @@ export async function copyExists(
   logger.info(`${sourceCommitHash} not found in ${destRepo}.`);
   return false;
 }
+
+interface CopyTask {
+  destRepo: string;
+  hash: string;
+}
+
+/**
+ * Searches for instances of the sourceCommitHash in recent pull requests and
+ * commits.
+ *
+ * @param sourceRepo the repo to copy commits from (org/repo).
+ * @param gitPath the local checkout of source repo, i.e., googelapis-gen.
+ * @param octokit an octokit instance
+ * @param maxWalk how far back should commits be walked?
+ */
+export async function runCopyAlgorithm(
+  sourceRepo: string,
+  gitPath: string,
+  database: ConfigsStore,
+  octokitArgs: OctokitParams,
+  maxWalk = 15
+) {
+  const octokit = await octokitFrom(octokitArgs);
+  const tasks: Array<CopyTask> = [];
+  let count = 0;
+  // Iterate over the maxWalk most recent commits to sourceRepo,
+  // checking for repos that require an update. The update required
+  // is pushed onto a stack, so that PRs are created in reverse
+  // chronological order:
+  //
+  // TODO(@bcoe): talk to rennie about the problem of commits that
+  // have no repos associated with them.
+  for await (const hash of commitsIterator(sourceRepo, octokit)) {
```

Comment:
@SurferJeffAtGoogle advantage of using API where we can, is it becomes easier to move back to something like Cloud Functions for some of our operations -- 6 of one half dozen of another if we're running in Cloud Build.

Reply:
SG.

Reply:
Wait, the commitsIterator needs an octokit to call the github API.  But the repo has already been cloned locally to disk, and the commit history can be observed there.  Why go back and call the github API?